src/backend/metal/src/command.rs

```diff
@@ -14,416 +14,112 @@
-                                                start);
-            }
-            Draw::IndexedInstanced(_count, _ninst, _start, _base, _offset) => unimplemented!(),
+        if self.should_restore {
+            self.encoder.restore_render_state();
+            self.should_restore = false;
         }
-
-        // self.cache.clear();
-        // self.buf.clear();
     }
 }
 
 impl command::Buffer<Resources> for CommandBuffer {
     fn reset(&mut self) {
-        self.cache.clear();
-        self.buf.clear();
+        self.encoder.reset();
     }
 
     fn bind_pipeline_state(&mut self, pso: Pipeline) {
-        self.buf.push(Command::BindPipeline(pso));
+        self.encoder.set_render_pipeline_state(pso.pipeline);
+        self.encoder.set_front_facing_winding(pso.winding);
+        self.encoder.set_cull_mode(pso.cull);
+        self.encoder.set_triangle_fill_mode(pso.fill);
+
+        // TODO(fkaa): do we need max value?
+        self.encoder.set_depth_bias(pso.depth_bias as f32 ,
+                                    pso.slope_scaled_depth_bias as f32,
+                                    0f32);
+
+        if let Some(depth_state) = pso.depth_stencil {
+            self.encoder.set_depth_stencil_state(depth_state);
+        }
     }
 
     fn bind_vertex_buffers(&mut self, vbs: pso::VertexBufferSet<Resources>) {
-        let mut buffers = [native::Buffer(MTLBuffer::nil()); MAX_VERTEX_ATTRIBUTES];
-        let mut offsets = [0; MAX_VERTEX_ATTRIBUTES];
-        let mut indices = [0; MAX_VERTEX_ATTRIBUTES];
-
-        for i in 0..1 {
+        for i in 0..MAX_VERTEX_ATTRIBUTES {
             if let Some((buffer, offset)) = vbs.0[i] {
-                buffers[i] = buffer.0;
-                offsets[i] = offset as u64;
-                indices[i] = i as u64;
+                // TODO(fkaa): is it okay if we cascade downwards with vertex
+                //             buffers? don't want to collide with user defined
+                //             values on accident!
+                self.encoder.set_vertex_buffer(30 - i as u64, offset as u64, unsafe { *(buffer.0).0 });
             }
         }
-        self.buf.push(Command::BindVertexBuffers(buffers, offsets, indices));
     }
 
     fn bind_constant_buffers(&mut self, cbs: &[pso::ConstantBufferParam<Resources>]) {
         for &stage in [Stage::Vertex, Stage::Pixel].iter() {
-            let mut buffers = [native::Buffer(MTLBuffer::nil()); MAX_CONSTANT_BUFFERS];
             let mask = stage.into();
-            let mut count = 0;
-            for cbuf in cbs.iter() {
-                if cbuf.1.contains(mask) {
-                    buffers[cbuf.2 as usize] = (cbuf.0).0;
-                    count += 1;
+            for cb in cbs.iter() {
+                if cb.1.contains(mask) {
+                    if let Stage::Vertex = stage {
+                        self.encoder.set_vertex_buffer(cb.2 as u64, 0, unsafe { *((cb.0).0).0 });
+                    } else {
```

Comment:
just to clarify: no geometry shaders, right? might be better to use a `match` here with some `unimplemented!` paths